### ISSUE TITLE
Critical styles missing during server render with Next.js

### ISSUE THREAD

**Issue by dcalhoun** [necolas/react-native-web#1079]
**The problem**
When server rendering a Next.js application, the "critical" styles for the given page are missing. They apply correctly, however, in the client.

**How to reproduce**
Simplified test case: https://github.com/HealthTeacher/nextjs-rnw

Steps to reproduce:
1.  Clone the test case.
1.  `yarn install`
1.  `yarn start`
1.  Open http://localhost:3000
1.  Disable JavaScript in your browser.
1.  Refresh the page.

**Expected behavior**
The "critical" styles for the page are rendered by the server.

**Environment (include versions). Did this work in previous versions?**

* React Native for Web (version): 0.8.9
* React (version): 16.4.2
* Browser: macOS Safari

## Notes:

- This seems related to [a previous `react-native-web` issue](https://github.com/necolas/react-native-web/issues/778#issuecomment-360385264), but I am failing to determine how to properly integrate Next.js as [the app renderer](https://github.com/necolas/react-native-web/issues/778#issuecomment-360704308).
- Next.js provides a way to define a [custom document](https://github.com/zeit/next.js#custom-document) that is often utilized integrate CSS-in-JS server rendering (e.g. [style-components](https://github.com/zeit/next.js/blob/canary/examples/with-styled-components/pages/_document.js#L7), [glamorous](https://github.com/zeit/next.js/blob/canary/examples/with-glamorous/pages/_document.js#L7), [emotion](https://github.com/zeit/next.js/blob/canary/examples/with-emotion/pages/_document.js#L7), [aphrodite](https://github.com/zeit/next.js/blob/canary/examples/with-aphrodite/pages/_document.js#L6)).
- Server rendered styles worked prior to the `react-native-web` [changing the default exports to ES6 modules in `0.7.0`](https://github.com/necolas/react-native-web/releases/tag/0.7.0).
- Using the `commonjs: true` for [`babel-plugin-react-native-web`](https://github.com/necolas/react-native-web/tree/master/packages/babel-plugin-react-native-web#usage) rather than compiling `react-native-web` with webpack fixes the missing styles (i.e. revert [this change](https://github.com/HealthTeacher/nextjs-rnw/commit/b0a9843c1b490f3a5d0ce6f7b0edff8626e24b8b)).

**Comment by steida** [comment]
I am using RNW here in Este and it seems it works. Maybe it can help.

https://github.com/este/este

**Comment by dcalhoun** [comment]
@steida thanks for the tip. Unfortunately, [you are utilizing the CommonJS version of RNW’s modules](https://github.com/este/este/blob/5491cf1158446e6866bf7a954b2e9ceada68ca86/.babelrc#L36), which I am already able to get working. 

My primary goal is to have server styles work when I compile RNW’s ES module exports with Babel.

**Comment by necolas** [comment]
Have you looked into why? What's different about the bundle?

**Comment by dcalhoun** [comment]
@necolas I have done a fair amount of debugging, but have been unable to identify what is going wrong. 

The only piece I’ve noticed and don’t quite understand is that, on the server, [`this._sheet` is always `null` here](https://github.com/necolas/react-native-web/blob/f062eded40a20464149f0675b1d22e236376b38f/packages/react-native-web/src/exports/StyleSheet/WebStyleSheet.js#L56) because it is [only set on the client](https://github.com/necolas/react-native-web/blob/f062eded40a20464149f0675b1d22e236376b38f/packages/react-native-web/src/exports/StyleSheet/WebStyleSheet.js#L35). That means any non-`initialStyles` _never_ gets “inserted” into the style sheet. *How/where do styles get inserted on the server?*

If you don’t have any specific thoughts or guidance, I’ll try to dive deeper and see if I can find something. I understand you can’t debug every integration/application. 

Thank you for your hard work on RNW.

**Comment by necolas** [comment]
Yeah the server only needs rules pushed into the array.

My first thought is that webpack is dropping some code. If the issue didn't exist for cjs understanding the difference between the 2 module formats is probably where I'd start

**Comment by dcalhoun** [comment]
@necolas below is the dependency tree arriving at `styleResolver.js` for Next.js' `page/index.js` (root route) and `pages/_document.js` ([shared wrapping component for each page](https://github.com/zeit/next.js/#custom-document)). During execution,`styleResolver.js` is imported first by `pages/index.js` and then later by `pages/_document.js`. The bolded `styleResolver.js` below (i.e. iii underneath `pages/_document.js`) is the focus of the issue. 

When using ES modules, a breakpoint in `styleResolver.js` is hit _twice_. When using CJS modules, the breakpoint is hit _once_. With CJS, it seems as though webpack caches the singleton `styleResolver` and does not import the file a second time. 

**Do you know why the `styleResolver.js` singleton executes twice for ES modules, but only once for CJS modules?**

- `pages/index.js`
  1. `Text`
  1. `TextStylePropTypes.js`
  1. `applyNativeProps.js`
  1. `NativeMethodsMixin.js`
  1. `createDOMProps.js`
  1. `styleResolver.js`
- `pages/_document.js`
  1. `AppRegistery.js`
  1. `renderApplication.js`
  1. **`styleResolver.js`** *- CJS never hits a debugger here, ES does.*

**Comment by necolas** [comment]
> Do you know why the styleResolver.js singleton executes twice for ES modules, but only once for CJS modules?

Nope. Can't tell if this is a webpack or Next.js thing either

**Comment by dcalhoun** [comment]
>Can't tell if this is a webpack or Next.js thing either

Alright. I agree this is likely not an issue with RNW. I'll close this issue and open an [issue in the Next.js community](https://spectrum.chat/thread/ba78d3a3-4ab3-48ef-9b77-956642dfbdb1).

Thank you for guidance.

**Comment by necolas** [comment]
Oh I don't know, it might be something to do with RNW. I don't mind leaving this open until we know what the issue is

**Comment by dcalhoun** [comment]
I was able to resolve this issue by [upgrading to `next@^7.0.0-canary.15`](https://github.com/HealthTeacher/nextjs-rnw/commit/ca5c8a086c688ef5a6b94d8c3d744c761a8771fb). The explanation of this issue may be found [here](https://bit.ly/2CJHzMM).

**TL;DR** – webpack doesn't use `require.cache` when compiling to the Node.js target, instead it uses it's own installed modules cache, which [caused issues for the way Next.js bundled pages](https://github.com/zeit/next.js/issues/4194#issuecomment-387443497).
